File: Combination Sum II.py
# leetcode 40 problem 
# 求数组中和为target的序列（单个序列中不允许重复取数）

# dfs实现
class Solution(object):
    def combinationSum2(self, candidates, target):
        res = []
        candidates.sort()
        self.dfs(candidates, target, 0, [], res)
        return res

    def dfs(self, nums, target, index, path, res):
        if target < 0:
            return  # backtracking
        if target == 0:
            if path not in res:
                res.append(path)
            return 
        if index > len(nums):
            return
        for i in range(index, len(nums)):
            self.dfs(nums, target-nums[i], i+1, path+[nums[i]], res)
        

File: Combination Sum II.java
// leetcode 40 problem
// 求数组中和为target的序列（不允许重复取数）

// 经典回溯，类比permutation、subset记忆
class Solution {
    public List<List<Integer>> combinationSum2(int[] nums, int target) {
        List<List<Integer>> list = new ArrayList<>();
        Arrays.sort(nums);
        backtrack(list, new ArrayList<>(), nums, target, 0);
        return list;

    }

    private void backtrack(List<List<Integer>> list, List<Integer> tempList, int [] nums, int remain, int start){
        if (remain < 0) return;
        else if (remain == 0) list.add(new ArrayList<>(tempList));
        else {
            for(int i = start; i < nums.length; i++){
                if(i > start && nums[i] == nums[i-1]) continue; // skip duplicates
                tempList.add(nums[i]);
                backtrack(list, tempList, nums, remain-nums[i], i + 1);
                tempList.remove(tempList.size() - 1); 
            }
        }
    } 
}